geode-core/src/main/java/org/apache/geode/distributed/internal/FunctionExecutionPooledExecutor.java

```diff
@@ -86,15 +96,32 @@ public void rejectedExecution(final Runnable r, ThreadPoolExecutor executor) {
             throw new RejectedExecutionException(
                 "executor has been shutdown");
           } else {
-            // System.out.println("Asif: Rejection called");
-            if (Thread
-                .currentThread() == ((FunctionExecutionPooledExecutor) executor).bufferConsumer) {
-              Thread th = executor.getThreadFactory().newThread((new Runnable() {
-                public void run() {
-                  r.run();
+            // The request was rejected due to all threads being in use.
+            //
+            // In the normal case, add the rejected request to the blocking queue.
+            //
+            // If the handler was invoked via the bufferConsumer (the thread that takes from the
+            // blocking queue and offers to the synchronous queue), then spin off a thread directly.
+            // In this case, an offer has already been made to the synchronous queue and failed, so
+            // all the function execution threads are in use.
+            //
+            // If the handler was invoked via a function execution thread, then spin off a thread
+            // directly. In this case, that means a function is executing another function. The
+            // child function request shouldn't be in the queue behind the parent request since the
+            // parent function is dependent on the child function executing.
+            final boolean isBufferConsumer = isBufferConsumer(executor);
+            if (isBufferConsumer || isFunctionExecutionThread()) {
```

Comment:
I think this code might be easier to understand if you get rid of this first if (isBufferConsumer || isFunctionExecutionThread) and just had "if (isBufferConsumer) ... else if (isFunctionExecThread) ... else".
The first two can both have calls to launchAdditionalThread.
I think it would help to split up the large comment into three; each in one block of this code.
It might even be worth extraction these three blocks into their own methods (then your comment can be on those methods)

Reply:
Thanks. I'll change this as you suggest.